File: client/src/lib/types/workflowMapper.ts
import type {
  WorkflowDataDto,
  WorkflowDto,
  EdgeDto,
  NodeDtoHeadersNodeDto,
  NodeDtoRequestNodeDto,
  NodeDto,
  SizeDto,
  PositionDto,
} from "$api";
import {
  type RequestNode,
  type HeadersNode,
  type Workflow,
  type WorkflowData,
  type WorkflowEdge,
  type WorkflowNode,
} from "./workflow";

export function toWorkflow(dto: WorkflowDto): Workflow {
  return {
    id: dto.id,
    name: dto.name,
    data: toWorkflowData(dto.data),
  };
}

export function toWorkflowData(dto: WorkflowDataDto): WorkflowData {
  return {
    nodes: dto.nodes?.map((x) => toWorkflowNode(x)) ?? [],
    edges: dto.edges?.map((x) => toWorkflowEdge(x)) ?? [],
  };
}

export function toWorkflowDataDto(data: WorkflowData): WorkflowDataDto {
  console.log(data);
  return {
    nodes: data.nodes?.map((x) => toWorkflowNodeDto(x)) ?? [],
    edges: data.edges?.map((x) => toWorkflowEdgeDto(x)) ?? [],
  };
}

function toWorkflowEdge(dto: EdgeDto): WorkflowEdge {
  return {
    id: dto.id,
    source: dto.source,
    target: dto.target,
  };
}

function toWorkflowEdgeDto(data: WorkflowEdge): EdgeDto {
  return {
    id: data.id,
    source: data.source,
    target: data.target,
  };
}

function toWorkflowNode(dto: NodeDto): WorkflowNode {
  const node = {
    ...dto,
    width: dto.size.width,
    height: dto.size.height,
    data: { ...dto }
  };

  switch (dto.type) {
    case "RequestNode":
      return node as RequestNode;
    case "HeadersNode":
    return node as HeadersNode;
    default:
      throw new Error(`Unknown NodeDto type '${dto.type}'`);
  };
}

function toWorkflowNodeDto(node: WorkflowNode): NodeDto {
  const size: SizeDto = { width: node.width, height: node.height };
  const position: PositionDto = node.position;
  const nodeData = node.data;
  const dto = { ...nodeData, size, position };

  switch (node.type) {
    case "RequestNode":
      return dto as NodeDtoRequestNodeDto;
    case "HeadersNode":
      return dto as NodeDtoHeadersNodeDto;
    default:
      throw new Error(`Unknown node type '${node.type}'`);
  };
}

File: server/Workflows/Api/Mappers/WorkflowMapper.cs
using WfAssist.Workflows.Api.Dtos;
using WfAssist.Workflows.Core.Models;

namespace WfAssist.Workflows.Api.Mappers;

internal static class WorkflowMapper
{
    internal static WorkflowDto ToDto(this Workflow entity)
        => new()
        {
            Id = entity.Id,
            Name = entity.Name,
            Data = entity.Data.ToDto()
        };

    internal static Workflow ToDomain(this WorkflowDto dto)
        => new(dto.Id, dto.Name, dto.Data.ToDomain());

    internal static WorkflowData ToDomain(this WorkflowDataDto dto)
        => new()
        {
            Edges = dto.Edges.Select(ToDomain),
            Nodes = dto.Nodes.Select(ToDomain)
        };

    private static WorkflowDataDto ToDto(this WorkflowData entity)
        => new()
        {
            Edges = entity.Edges.Select(ToDto),
            Nodes = entity.Nodes.Select(ToDto)
        };

    private static EdgeDto ToDto(this Edge entity)
        => new()
        {
            Id = entity.Id,
            Source = entity.Source,
            Target = entity.Target,
        };

    private static Edge ToDomain(this EdgeDto dto)
        => new()
        {
            Id = dto.Id,
            Source = dto.Source,
            Target = dto.Target,
        };

    private static NodeDto ToDto(this Node entity)
    {
        return entity switch
        {
            RequestNode requestNode => new RequestNodeDto
            {
                Id = requestNode.Id,
                Size = new SizeDto(requestNode.Size.Width, requestNode.Size.Height),
                Position = new PositionDto(requestNode.Position.X, requestNode.Position.Y),
                RequestType = requestNode.RequestType.ToDto(),
                Url = requestNode.Url,
                RequestBody = requestNode.RequestBody
            },
            HeadersNode headersNode => new HeadersNodeDto
            {
                Id = headersNode.Id,
                Size = new SizeDto(headersNode.Size.Width, headersNode.Size.Height),
                Position = new PositionDto(headersNode.Position.X, headersNode.Position.Y),
                Headers = headersNode.Headers.Select(x => new HttpHeaderDto(x.Name, x.Value)).ToList()
            },
            _ => throw new ArgumentOutOfRangeException(nameof(entity))
        };
    }

    private static Node ToDomain(this NodeDto dto)
    {
        return dto switch
        {
            RequestNodeDto requestNodeDto => new RequestNode
            {
                Id = requestNodeDto.Id,
                Size = new Size(requestNodeDto.Size.Width, requestNodeDto.Size.Height),
                Position = new Position(requestNodeDto.Position.X, requestNodeDto.Position.Y),
                RequestType = requestNodeDto.RequestType.ToDomain(),
                Url = requestNodeDto.Url,
                RequestBody = requestNodeDto.RequestBody
            },
            HeadersNodeDto headersNodeDto => new HeadersNode
            {
                Id = headersNodeDto.Id,
                Size = new Size(headersNodeDto.Size.Width, headersNodeDto.Size.Height),
                Position = new Position(headersNodeDto.Position.X, headersNodeDto.Position.Y),
                Headers = headersNodeDto.Headers.Select(x => new HttpHeader(x.Name, x.Value)).ToList()
            },
            _ => throw new ArgumentOutOfRangeException(nameof(dto))
        };
    }

    private static RequestType ToDomain(this RequestTypeDto dto)
    {
        return dto switch
        {
            RequestTypeDto.Get => RequestType.Get,
            RequestTypeDto.Post => RequestType.Post,
            RequestTypeDto.Put => RequestType.Put,
            RequestTypeDto.Patch => RequestType.Patch,
            RequestTypeDto.Delete => RequestType.Delete,
            _ => throw new InvalidOperationException($"Unknown RequestType type {dto}")
        };
    }

    private static RequestTypeDto ToDto(this RequestType requestType)
    {
        return requestType switch
        {
            RequestType.Get => RequestTypeDto.Get,
            RequestType.Post => RequestTypeDto.Post,
            RequestType.Put => RequestTypeDto.Put,
            RequestType.Patch => RequestTypeDto.Patch,
            RequestType.Delete => RequestTypeDto.Delete,
            _ => throw new InvalidOperationException($"Unknown RequestType type {requestType}")
        };
    }
}
